src/util/apparmor.c

```diff
@@ -93,3 +105,446 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
+        int allow = false, audit = true;
+
+        if (!is_apparmor_enabled(registry))
+                return 0;
+
+        if (!sender_context_dup || !receiver_context_dup)
+                return -ENOMEM;
+
+        sender_security_label = aa_splitcon(sender_context_dup, &sender_security_mode);
+        receiver_security_label = aa_splitcon(receiver_context_dup, &receiver_security_mode);
+
+        if (is_unconfined(sender_security_label) && is_unconfined(receiver_security_label))
+                return 0;
+
+        r = apparmor_message_query(check_send,
+                                   check_send ? sender_security_label : receiver_security_label,
+                                   registry->bustype,
+                                   check_send ? receiver_security_label : sender_security_label,
+                                   subject, subject_id, path, interface, method, &allow, &audit);
+        if (r < 0)
+                return error_origin(r);
+
+        if (audit)
+                bus_apparmor_log("apparmor=\"%s\" operation=\"dbus_%s\" bus=\"%s\" path=\"%s\" interface=\"%s\" method=\"%s\"", allowstr(allow), direction, registry->bustype, path, interface, method);
+
+        if (is_complain(check_send ? sender_security_mode : receiver_security_mode))
+                allow = 1;
+
+        if (!allow)
+                return BUS_APPARMOR_E_DENIED;
+
+        return 0;
+}
+
+/**
+ * bus_apparmor_check_eavesdrop() - check if the given context may eavesdrop
+ * @registry:           AppArmor registry to operate on
+ * @context:            security context that wants to eavesdrop
+ *
+ * Check if the given sender context is allowed to do eavesdropping.
+ *
+ * Return: 0 if the transaction is allowed, BUS_APPARMOR_E_DENIED if it is not,
+ *         or a negative error code on failure.
+ */
+int bus_apparmor_check_eavesdrop(BusAppArmorRegistry *registry,
+                                 const char *context)
+{
+        _c_cleanup_(c_freep) char *condup = strdup(context);
+        char *security_label, *security_mode;
+        _c_cleanup_(c_freep) char *qstr = NULL;
+        int r;
+        /* the AppArmor API uses pointers to int for pointers to boolean */
+        int allow = false, audit = true;
+
+        if (!is_apparmor_enabled(registry) || is_unconfined(context))
+                return 0;
+
+        if (!condup)
+                return -ENOMEM;
+
+        security_label = aa_splitcon(condup, &security_mode);
+
+        r = build_eavesdrop_query(&qstr, security_label, registry->bustype);
+        if (r < 0)
+                return error_origin(r);
+
+        r = aa_query_label(AA_DBUS_EAVESDROP, qstr, r, &allow, &audit);
+        if (r < 0)
+                return error_origin(-errno);
+
+        if (audit)
+                bus_apparmor_log("apparmor=\"%s\" operation=\"dbus_eavesdrop\" bus=\"%s\" label=\"%s\"", allowstr(allow), registry->bustype, context);
+
+        if (is_complain(security_mode))
+                allow = 1;
```

Comment:
A similar question here, should this be moved above the logging line, so the string matches the result?